pkg/controller/core/workload_controller.go

```diff
@@ -103,10 +143,29 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 				"Inadmissible", fmt.Sprintf("ClusterQueue %s is inactive", cqName))
 			return ctrl.Result{}, client.IgnoreNotFound(err)
 		}
-	case admitted:
-		msg := fmt.Sprintf("Admitted by ClusterQueue %s", wl.Spec.Admission.ClusterQueue)
-		err := workload.UpdateStatusIfChanged(ctx, r.client, &wl, kueue.WorkloadAdmitted, metav1.ConditionTrue, "AdmissionByKueue", msg)
+	case cancellingAdmission:
+		err := workload.UpdateStatusIfChanged(ctx, r.client, &wl, kueue.WorkloadAdmitted, metav1.ConditionFalse,
+			"AdmissionCancelled", "Admission cancelled")
 		return ctrl.Result{}, client.IgnoreNotFound(err)
+	case admitted:
+		if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadAdmitted) {
+			countingTowardsTimeout, recheckAfter := r.admittedNotReadyWorkload(&wl, realClock)
+			if countingTowardsTimeout {
+				if recheckAfter > 0 {
+					klog.V(4).InfoS("Workload not yet ready and did not exceed its timeout", "workload", req.NamespacedName.String(), "recheckAfter", recheckAfter)
+					return ctrl.Result{RequeueAfter: recheckAfter}, nil
+				} else {
+					klog.V(3).InfoS("Cancelling admission of the workload due to exceeding the PodsReady timeout", "workload", req.NamespacedName.String())
+					wl.Spec.Admission = nil
+					err := r.client.Patch(ctx, workload.AdmissionPatch(&wl), client.Apply, client.FieldOwner(constants.AdmissionName))
+					return ctrl.Result{}, client.IgnoreNotFound(err)
+				}
+			}
```

Comment:
I think it's worth having all of this in its own method to reduce indentation.

Then it's probably not worth having a separate method just for `admittedNotReadyWorkload`.

Reply:
If for the selected code, should we name it like `handlePodsReadyTimeout`? Maybe a new function `reconcileAdmittedWorkload` for the entire case?

Reply:
Extracted to a new function, but for now I still keep `admittedNotReadyWorkload` because it has unit tests added and it seems to have a nicer interface for unit testing (no need for context or request)